qiskit/providers/basic_provider/basic_simulator.py

```diff
@@ -0,0 +1,747 @@
+
+The simulator is run using
+
+.. code-block:: python
+
+   BasicSimulator().run(run_input)
+
+Where the input is a :class:`.QuantumCircuit` object and the
+output is a :class:`.BasicProviderJob` object,
+which can later be queried for the Result object. The result will contain a 'memory' data
+field, which is a result of measurements for each shot.
+"""
+
+from __future__ import annotations
+
+import uuid
+import time
+import logging
+import warnings
+
+from collections import Counter
+import numpy as np
+
+from qiskit.circuit import QuantumCircuit
+from qiskit.circuit.library import UnitaryGate
+from qiskit.circuit.library.standard_gates import get_standard_gate_name_mapping
+from qiskit.providers import Provider
+from qiskit.providers.backend import BackendV2
+from qiskit.providers.models import BackendConfiguration
+from qiskit.providers.options import Options
+from qiskit.qobj import QasmQobj, QasmQobjConfig, QasmQobjExperiment
+from qiskit.result import Result
+from qiskit.transpiler import Target
+
+from .basic_provider_job import BasicProviderJob
+from .basic_provider_tools import single_gate_matrix
+from .basic_provider_tools import SINGLE_QUBIT_GATES
+from .basic_provider_tools import cx_gate_matrix
+from .basic_provider_tools import einsum_vecmul_index
+from .exceptions import BasicProviderError
+
+logger = logging.getLogger(__name__)
+
+
+class BasicSimulator(BackendV2):
+    """Python implementation of a basic (non-efficient) quantum simulator.
+
+    This implementation was originally based on the :class:`.BackendV1` interface,
+    and later migrated to follow :class:`.BackendV2`.
+    """
+
+    # Formerly calculated as `int(log2(local_hardware_info()["memory"]*(1024**3)/16))`.
+    # After the removal of `local_hardware_info()`, it's hardcoded to 24 qubits,
+    # which matches the ~268 MB of required memory.
+    MAX_QUBITS_MEMORY = 24
+
+    def __init__(
+        self,
+        provider: Provider | None = None,
+        target: Target | None = None,
+        **fields,
+    ) -> None:
+        """
+        Args:
+            provider: An optional backwards reference to the
+                :class:`~qiskit.providers.Provider` object that the backend
+                is from.
+            target: An optional target to configure the simulator.
+            fields: kwargs for the values to use to override the default
+                options.
+
+        Raises:
+            AttributeError: If a field is specified that's outside the backend's
+                options.
+        """
+
+        super().__init__(
+            provider=provider,
+            name="basic_simulator",
+            description="A python simulator for quantum experiments",
+            backend_version="0.1",
+            **fields,
+        )
+
+        self._target = target
+        self._configuration = None
+
+        # Internal simulator variables
+        self._local_random = np.random.RandomState()
```

Comment:
This is actually the legacy interface for numpy random number generators. We probably should use the newer interface: https://numpy.org/doc/stable/reference/random/generator.html there are some small changes needed but it shouldn't be too bad.

Reply:
I was not sure about making this change because the number generation algorithm is different with the new interface. At least, I have not found a way to generate the same values with the same seed (I noticed this when migrating `algorithm_globals`). But I keep forgetting that this is technically a new class, so that is technically not a behavior change.